### DISCUSSION TITLE
Questions on Micropython and ulab library (Numpy)

### DISCUSSION THREAD

**Discussion by kovalick** [v923z/micropython-ulab#682] [General]
I am using the Thonny editor (Windows 11) attached to a Rasp Pi Pico running Micropython. Works perfect so far. I need to add the ulab library (Numpy alternative for Micropython). The ulab library install directions are located at https://github.com/v923z/micropython-ulab/blob/master/README.md#rp2-based-boards. 

I need some advice for implementing the suggested Bash script to compile the ulab library.  I am using Windows 11 and it wont run a Bash script natively. Do I need to install the Windows subsystem for Linux to run the script? Given the popularity of ulab, is there a "premade" ulab library that I can install without compiling the ulab C code? Is there a version of Micropython with ulab included? 

Any advice will help..  Thanks   Al (San Francisco)

**Comment by v923z** [comment]
If you want to compile the firmware yourself, there are a couple of options:

1. As you've pointed out, you can install linux on a virtual machine, and then run the script there.
2. You can also use github for this. All you have to do is set your compilation platform, and then create a release, so that the firmware is not deleted after the github runner is done. You can use this as a template: https://github.com/v923z/micropython-builder. 

Which brings me to my final point: the repository above already has pre-compiled firmware. You simply download the binary, and upload it to your MCU via thonny.

I hope this helps.

**Reply by kovalick** [reply]
This is very useful advice! I really appreciate your effort to assist me on this. I did not see the precompiled firmware and was wondering why not. Are you saying there is a "firmware.rp2" file pre compiled (with ulab) and ready for Pico upload? If so, can you please provide the link? Thanks!  Al

**Comment by v923z** [comment]
Here are the firmware files: https://github.com/v923z/micropython-builder/releases

**Comment by kovalick** [comment]
Hello
 Got it. I had missed this before since it's not on the main GitHub ulab page.  Thanks for your hard work on ulab. I am using it to decode signals with multiple tones.   Al
________________________________
